protobuf-maven-plugin/src/main/java/io/github/ascopes/protobufmavenplugin/plugin/BinaryPluginResolver.java

```diff
@@ -104,29 +106,40 @@ private ResolvedProtocPlugin resolveMavenPlugin(
         .next();
 
     makeExecutable(path);
-    return createResolvedProtocPlugin(plugin, path);
+    var resolvedPlugin = createResolvedProtocPlugin(plugin, path);
+    return Optional.of(resolvedPlugin);
   }
 
-  private ResolvedProtocPlugin resolvePathPlugin(
+  private Optional<ResolvedProtocPlugin> resolvePathPlugin(
       PathProtocPlugin plugin
   ) throws ResolutionException {
-    var path = systemPathResolver.resolve(plugin.getName())
-        .orElseThrow(() -> new ResolutionException(
-            "No executable '"
-                + plugin.getName()
-                + "' was found on the system path"
-        ));
-
-    return createResolvedProtocPlugin(plugin, path);
+    var maybePath = systemPathResolver.resolve(plugin.getName());
+
+    if (maybePath.isEmpty() && plugin.isOptional()) {
+      return Optional.empty();
+    }
+
+    var path = maybePath.orElseThrow(() -> new ResolutionException(
+        "No plugin named " + plugin.getName() + " was found on the system path"
+    ));
+
+    return Optional.of(createResolvedProtocPlugin(plugin, path));
   }
 
-  private ResolvedProtocPlugin resolveUrlPlugin(
+  private Optional<ResolvedProtocPlugin> resolveUrlPlugin(
       UrlProtocPlugin plugin
   ) throws ResolutionException {
     var path = urlResourceFetcher.fetchFileFromUrl(plugin.getUrl(), ".exe");
+
+    // TODO: handle path being missing consistently.
```

Comment:
Right now the URL handler appears to emit a missing path if a file:// is provided that does not exist, since we're relying on streaming to fail for non-file URLs to signal that they do not exist. We skip that on file:// URLs as nothing needs to be streamed.

Will fix this in this PR. Planning to change the URL resolver to emit optionals if resources are not found.

I may also merge the Http and UrlConnection behaviours together.